**EXTRACT (COLLECTING THE DATA)**

**Importing Libraries**

In [1]:
# Cell 1: Import the tools (libraries) we need
import requests      # This tool allows Python to 'talk' to the API website
import pandas as pd  # This tool allows us to organize data into tables
from datetime import datetime # This tool helps us get the current date and time

print("Tools imported successfully. Ready for Step 2!")

Tools imported successfully. Ready for Step 2!


**Setting the API Key**

In [2]:
# To set API key here
API_KEY = "YOUR-API-KEY"

print("API Key saved!")

API Key saved!


**Defining Cities and Base URL**

In [3]:
# Set the cities to track
# We have included Port Harcourt, Accra, and Cape Town
cities = ["Port Harcourt", "Accra", "Cape Town"]

# This is the web address to get the data
base_url = "http://api.openweathermap.org/data/2.5/weather"

# This is an empty list to store the data later
weather_data = []

print("Cities list (Port Harcourt, Accra, Cape Town) is ready!")

Cities list (Port Harcourt, Accra, Cape Town) is ready!


**The Collection Loop**

In [4]:
# Create the collection loop
for city in cities:
    params = {"q": city, "appid": API_KEY, "units": "metric"}
    response = requests.get(base_url, params=params)

    if response.status_code == 200:
        data = response.json()
        weather_data.append({
            "City": data["name"],
            "Temperature": data["main"]["temp"],
            "Humidity": data["main"]["humidity"],
            "Condition": data["weather"][0]["description"],
            "Wind_Speed": data["wind"]["speed"],
            "Date_Time": datetime.now().strftime("%Y-%m-%d %H:%M:%S")
        })
        print(f"Collected: {city}")

Collected: Port Harcourt
Collected: Accra
Collected: Cape Town


**TRANSFORM DATA (PROCESSING THE DATA)**

**Creating the Data Table**

In [5]:
# Show the results
print(weather_data)

[{'City': 'Port Harcourt', 'Temperature': 24.32, 'Humidity': 90, 'Condition': 'light rain', 'Wind_Speed': 2.58, 'Date_Time': '2026-07-15 22:23:44'}, {'City': 'Accra', 'Temperature': 24.3, 'Humidity': 85, 'Condition': 'broken clouds', 'Wind_Speed': 5.03, 'Date_Time': '2026-07-15 22:23:44'}, {'City': 'Cape Town', 'Temperature': 13.33, 'Humidity': 91, 'Condition': 'broken clouds', 'Wind_Speed': 0.72, 'Date_Time': '2026-07-15 22:23:44'}]


**Displaying the Data for Comparison**

In [6]:
# Create the table
df = pd.DataFrame(weather_data)

# Show the clean table
print(df)

            City  Temperature  Humidity      Condition  Wind_Speed  \
0  Port Harcourt        24.32        90     light rain        2.58   
1          Accra        24.30        85  broken clouds        5.03   
2      Cape Town        13.33        91  broken clouds        0.72   

             Date_Time  
0  2026-07-15 22:23:44  
1  2026-07-15 22:23:44  
2  2026-07-15 22:23:44  


In [7]:
# Displaying the data for comparison
# Looking at relevant fields to compare the three cities
print(df[['City', 'Temperature', 'Humidity', 'Condition']])

            City  Temperature  Humidity      Condition
0  Port Harcourt        24.32        90     light rain
1          Accra        24.30        85  broken clouds
2      Cape Town        13.33        91  broken clouds


**Identifying Insights**

In [8]:
# Identifying the city with the highest humidity
# Comparing humidity values to determine which city is the most humid
max_humidity_row = df.loc[df['Humidity'].idxmax()]
print(f"City with highest humidity: {max_humidity_row['City']} at {max_humidity_row['Humidity']}%")

City with highest humidity: Cape Town at 91%


**LOAD (STORING THE DATA)**

In [9]:
# Task 3: Load Data
# Save the table as a CSV file
df.to_csv("weather_data.csv", index=False)

print("Data saved successfully as 'weather_data.csv'!")

Data saved successfully as 'weather_data.csv'!


**KEY INSIGHTS**


*   Temperature: Port Harcourt and Accra have similar temperatures at 24.32°C and 24.30°C, while Cape Town is much cooler at 13.33°C.

*   Humidity: After comparing the humidity values, I found that Cape Town has the highest humidity at 91%. Port Harcourt follows with 90%, while Accra has the lowest humidity at 85%.

*   Weather Conditions: Port Harcourt is experiencing light rain, while Accra and Cape Town are both experiencing broken clouds. This shows that the weather conditions vary across the three cities.





**CONCLUSION**

This project successfully demonstrates the effectiveness of an ETL (Extract, Transform, Load) pipeline for real-time data analysis. By automating the extraction of weather data via the OpenWeather API, I was able to transform raw, unstructured responses into a clean, organized Pandas DataFrame, which allowed for efficient comparison and analysis. Storing this information as a CSV file provides a reliable foundation for future tracking or long-term trend analysis. Ultimately, this pipeline shows how organizations can move from raw data collection to actionable insights in a structured and reproducible way.